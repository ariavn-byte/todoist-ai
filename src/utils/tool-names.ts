/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */
export const ToolNames = {
    // Task management tools
    ADD_TASKS: 'add-tasks',
    COMPLETE_TASKS: 'complete-tasks',
    UPDATE_TASKS: 'update-tasks',
    FIND_TASKS: 'find-tasks',
    FIND_TASKS_BY_DATE: 'find-tasks-by-date',
    FIND_COMPLETED_TASKS: 'find-completed-tasks',

    // Project management tools
    ADD_PROJECTS: 'add-projects',
    UPDATE_PROJECTS: 'update-projects',
    FIND_PROJECTS: 'find-projects',

    // Section management tools
    ADD_SECTIONS: 'add-sections',
    UPDATE_SECTIONS: 'update-sections',
    FIND_SECTIONS: 'find-sections',

    // General tools
    GET_OVERVIEW: 'get-overview',
    DELETE_OBJECT: 'delete-object',
} as const

// Type for all tool names
export type ToolName = (typeof ToolNames)[keyof typeof ToolNames]
