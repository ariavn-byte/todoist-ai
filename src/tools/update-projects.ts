import type { PersonalProject, WorkspaceProject } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { FIND_PROJECTS, FIND_TASKS, GET_OVERVIEW } = ToolNames

const ProjectUpdateSchema = z.object({
    id: z.string().min(1).describe('The ID of the project to update.'),
    name: z.string().min(1).optional().describe('The new name of the project.'),
    isFavorite: z.boolean().optional().describe('Whether the project is a favorite.'),
    viewStyle: z.enum(['list', 'board', 'calendar']).optional().describe('The project view style.'),
})

type ProjectUpdate = z.infer<typeof ProjectUpdateSchema>

const ArgsSchema = {
    projects: z.array(ProjectUpdateSchema).min(1).describe('The projects to update.'),
}

const updateProjects = {
    name: ToolNames.UPDATE_PROJECTS,
    description: 'Update multiple existing projects with new values.',
    parameters: ArgsSchema,
    async execute(args, client) {
        const { projects } = args
        const updateProjectsPromises = projects.map(async (project) => {
            if (!hasUpdatesToMake(project)) {
                return undefined
            }

            const { id, ...updateArgs } = project
            return await client.updateProject(id, updateArgs)
        })

        const updatedProjects = (await Promise.all(updateProjectsPromises)).filter(
            (project): project is PersonalProject | WorkspaceProject => project !== undefined,
        )

        const textContent = generateTextContent({
            projects: updatedProjects,
            args,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                projects: updatedProjects,
                totalCount: updatedProjects.length,
                updatedProjectIds: updatedProjects.map((project) => project.id),
                appliedOperations: {
                    updateCount: updatedProjects.length,
                    skippedCount: projects.length - updatedProjects.length,
                },
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    projects,
    args,
}: {
    projects: (PersonalProject | WorkspaceProject)[]
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
}) {
    const totalRequested = args.projects.length
    const actuallyUpdated = projects.length
    const skipped = totalRequested - actuallyUpdated

    const count = projects.length
    const projectList = projects.map((project) => `â€¢ ${project.name} (id=${project.id})`).join('\n')

    let summary = `Updated ${count} project${count === 1 ? '' : 's'}`
    if (skipped > 0) {
        summary += ` (${skipped} skipped - no changes)`
    }

    if (count > 0) {
        summary += `:\n${projectList}`
    }

    // Context-aware next steps for updated projects
    const nextSteps: string[] = []

    if (projects.length > 0) {
        if (count === 1) {
            const project = projects[0]
            if (project) {
                nextSteps.push(
                    `Use ${GET_OVERVIEW} with projectId=${project.id} to see project structure`,
                )
                nextSteps.push(
                    `Use ${FIND_TASKS} with projectId=${project.id} to review existing tasks`,
                )
            }
        } else {
            nextSteps.push(`Use ${FIND_PROJECTS} to see all projects with updated names`)
            nextSteps.push(`Use ${GET_OVERVIEW} to see updated project hierarchy`)
        }
    } else {
        nextSteps.push(`Use ${FIND_PROJECTS} to see current projects`)
    }

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

function hasUpdatesToMake({ id, ...otherUpdateArgs }: ProjectUpdate) {
    return Object.keys(otherUpdateArgs).length > 0
}

export { updateProjects }
