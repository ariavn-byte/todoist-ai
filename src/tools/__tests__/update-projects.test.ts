import type { PersonalProject, TodoistApi, WorkspaceProject } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import {
    createMockProject,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { updateProjects } from '../update-projects.js'

// Mock the Todoist API
const mockTodoistApi = {
    updateProject: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { FIND_PROJECTS, UPDATE_PROJECTS, GET_OVERVIEW } = ToolNames

describe(`${UPDATE_PROJECTS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('updating a single project', () => {
        it('should update a project when id and name are provided', async () => {
            const mockApiResponse: PersonalProject = {
                url: 'https://todoist.com/projects/existing-project-123',
                id: 'existing-project-123',
                parentId: null,
                isDeleted: false,
                updatedAt: '2025-08-13T22:10:30.000000Z',
                childOrder: 1,
                description: '',
                isCollapsed: false,
                canAssignTasks: false,
                color: 'red',
                isFavorite: false,
                isFrozen: false,
                name: 'Updated Project Name',
                viewStyle: 'list',
                isArchived: false,
                inboxProject: false,
                isShared: false,
                createdAt: '2024-01-01T00:00:00Z',
                defaultOrder: 0,
            }

            mockTodoistApi.updateProject.mockResolvedValue(mockApiResponse)

            const result = await updateProjects.execute(
                { projects: [{ id: 'existing-project-123', name: 'Updated Project Name' }] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.updateProject).toHaveBeenCalledWith('existing-project-123', {
                name: 'Updated Project Name',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated 1 project:')
            expect(textContent).toContain('Updated Project Name (id=existing-project-123)')
            expect(textContent).toContain(`Use ${GET_OVERVIEW} with projectId=existing-project-123`)

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    projects: [mockApiResponse],
                    totalCount: 1,
                    updatedProjectIds: ['existing-project-123'],
                    appliedOperations: {
                        updateCount: 1,
                        skippedCount: 0,
                    },
                }),
            )
        })

        it('should update project with isFavorite and viewStyle options', async () => {
            const mockApiResponse: PersonalProject = {
                url: 'https://todoist.com/projects/project-123',
                id: 'project-123',
                parentId: null,
                isDeleted: false,
                updatedAt: '2025-08-13T22:10:30.000000Z',
                childOrder: 1,
                description: '',
                isCollapsed: false,
                canAssignTasks: false,
                color: 'red',
                isFavorite: true,
                isFrozen: false,
                name: 'Updated Favorite Project',
                viewStyle: 'board',
                isArchived: false,
                inboxProject: false,
                isShared: false,
                createdAt: '2024-01-01T00:00:00Z',
                defaultOrder: 0,
            }

            mockTodoistApi.updateProject.mockResolvedValue(mockApiResponse)

            const result = await updateProjects.execute(
                {
                    projects: [
                        {
                            id: 'project-123',
                            name: 'Updated Favorite Project',
                            isFavorite: true,
                            viewStyle: 'board',
                        },
                    ],
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.updateProject).toHaveBeenCalledWith('project-123', {
                name: 'Updated Favorite Project',
                isFavorite: true,
                viewStyle: 'board',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated 1 project:')
            expect(textContent).toContain('Updated Favorite Project (id=project-123)')
        })
    })

    describe('updating multiple projects', () => {
        it('should update multiple projects and return mapped results', async () => {
            type Project = PersonalProject | WorkspaceProject
            const mockProjects: [Project, Project, Project] = [
                createMockProject({ id: 'project-1', name: 'Updated First Project' }),
                createMockProject({ id: 'project-2', name: 'Updated Second Project' }),
                createMockProject({ id: 'project-3', name: 'Updated Third Project' }),
            ]

            const [project1, project2, project3] = mockProjects
            mockTodoistApi.updateProject
                .mockResolvedValueOnce(project1)
                .mockResolvedValueOnce(project2)
                .mockResolvedValueOnce(project3)

            const result = await updateProjects.execute(
                {
                    projects: [
                        { id: 'project-1', name: 'Updated First Project' },
                        { id: 'project-2', name: 'Updated Second Project' },
                        { id: 'project-3', name: 'Updated Third Project' },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly for each project
            expect(mockTodoistApi.updateProject).toHaveBeenCalledTimes(3)
            expect(mockTodoistApi.updateProject).toHaveBeenNthCalledWith(1, 'project-1', {
                name: 'Updated First Project',
            })
            expect(mockTodoistApi.updateProject).toHaveBeenNthCalledWith(2, 'project-2', {
                name: 'Updated Second Project',
            })
            expect(mockTodoistApi.updateProject).toHaveBeenNthCalledWith(3, 'project-3', {
                name: 'Updated Third Project',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated 3 projects:')
            expect(textContent).toContain('Updated First Project (id=project-1)')
            expect(textContent).toContain('Updated Second Project (id=project-2)')
            expect(textContent).toContain('Updated Third Project (id=project-3)')
            expect(textContent).toContain(`Use ${FIND_PROJECTS} to see all projects`)

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    projects: mockProjects,
                    totalCount: 3,
                    updatedProjectIds: ['project-1', 'project-2', 'project-3'],
                    appliedOperations: {
                        updateCount: 3,
                        skippedCount: 0,
                    },
                }),
            )
        })

        it('should skip projects with no updates and report correctly', async () => {
            const mockProject = createMockProject({
                id: 'project-1',
                name: 'Updated Project',
            })

            mockTodoistApi.updateProject.mockResolvedValue(mockProject)

            const result = await updateProjects.execute(
                {
                    projects: [
                        { id: 'project-1', name: 'Updated Project' },
                        { id: 'project-2' }, // No name provided, should be skipped
                    ],
                },
                mockTodoistApi,
            )

            // Should only call API once for the project with actual updates
            expect(mockTodoistApi.updateProject).toHaveBeenCalledTimes(1)
            expect(mockTodoistApi.updateProject).toHaveBeenCalledWith('project-1', {
                name: 'Updated Project',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Updated 1 project (1 skipped - no changes):')
            expect(textContent).toContain('Updated Project (id=project-1)')

            // Verify structured content reflects skipped count
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    appliedOperations: {
                        updateCount: 1,
                        skippedCount: 1,
                    },
                }),
            )
        })
    })

    describe('error handling', () => {
        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Project not found')
            mockTodoistApi.updateProject.mockRejectedValue(apiError)

            await expect(
                updateProjects.execute(
                    { projects: [{ id: 'nonexistent', name: 'New Name' }] },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Project not found')
        })

        it('should handle partial failures in multiple projects', async () => {
            const mockProject = createMockProject({
                id: 'project-1',
                name: 'Updated Project',
            })

            mockTodoistApi.updateProject
                .mockResolvedValueOnce(mockProject)
                .mockRejectedValueOnce(new Error('API Error: Project not found'))

            await expect(
                updateProjects.execute(
                    {
                        projects: [
                            { id: 'project-1', name: 'Updated Project' },
                            { id: 'nonexistent', name: 'New Name' },
                        ],
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Project not found')
        })
    })
})
