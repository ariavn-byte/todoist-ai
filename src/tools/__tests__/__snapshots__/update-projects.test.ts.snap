// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`update-projects tool updating a single project should update a project when id and name are provided 1`] = `
"Updated 1 project:
• Updated Project Name (id=existing-project-123)
Next:
- Use get-overview with projectId=existing-project-123 to see project structure
- Use find-tasks with projectId=existing-project-123 to review existing tasks"
`;

exports[`update-projects tool updating a single project should update project with isFavorite and viewStyle options 1`] = `
"Updated 1 project:
• Updated Favorite Project (id=project-123)
Next:
- Use get-overview with projectId=project-123 to see project structure
- Use find-tasks with projectId=project-123 to review existing tasks"
`;

exports[`update-projects tool updating multiple projects should skip projects with no updates and report correctly 1`] = `
"Updated 1 project (1 skipped - no changes):
• Updated Project (id=project-1)
Next:
- Use get-overview with projectId=project-1 to see project structure
- Use find-tasks with projectId=project-1 to review existing tasks"
`;

exports[`update-projects tool updating multiple projects should update multiple projects and return mapped results 1`] = `
"Updated 3 projects:
• Updated First Project (id=project-1)
• Updated Second Project (id=project-2)
• Updated Third Project (id=project-3)
Next:
- Use find-projects to see all projects with updated names
- Use get-overview to see updated project hierarchy"
`;
